compiler/src/language_server/completion.re

```diff
@@ -0,0 +1,547 @@
+let completion_item_kind_of_yojson = json =>
+  Result.bind(json |> [%of_yojson: int], value => {
+    switch (completion_item_kind_of_enum(value)) {
+    | Some(severity) => Ok(severity)
+    | None => Result.Error("Invalid enum value")
+    }
+  });
+
+let completion_trigger_kind_to_yojson = kind =>
+  completion_trigger_kind_to_enum(kind) |> [%to_yojson: int];
+let completion_trigger_kind_of_yojson = json =>
+  Result.bind(json |> [%of_yojson: int], value => {
+    switch (completion_trigger_kind_of_enum(value)) {
+    | Some(kind) => Ok(kind)
+    | None => Result.Error("Invalid enum value")
+    }
+  });
+
+[@deriving yojson]
+type completion_item = {
+  label: string,
+  kind: completion_item_kind,
+  detail: string,
+  documentation: string,
+};
+
+[@deriving yojson({strict: false})]
+type completion_context = {
+  [@key "triggerKind"]
+  trigger_kind: completion_trigger_kind,
+  [@key "triggerCharacter"] [@default None]
+  trigger_character: option(string),
+};
+
+// https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#completionParams
+module RequestParams = {
+  [@deriving yojson({strict: false})]
+  type t = {
+    [@key "textDocument"]
+    text_document: Protocol.text_document_identifier,
+    position: Protocol.position,
+    [@default None]
+    context: option(completion_context),
+  };
+};
+
+// https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#completionList
+module ResponseResult = {
+  [@deriving yojson]
+  type t = {
+    isIncomplete: bool,
+    items: list(completion_item),
+  };
+};
+
+let send_completion =
+    (~id: Protocol.message_id, completions: list(completion_item)) => {
+  Protocol.response(
+    ~id,
+    ResponseResult.to_yojson({isIncomplete: false, items: completions}),
+  );
+};
+
+type completionState =
+  | CompletableCode(string)
+  | CompletableExpr(string)
+  | CompletableType(string)
+  | CompletableInclude(string);
+
+let convert_position_to_offset = (source, position: Protocol.position) => {
+  let lines = String.split_on_char('\n', source);
+  List_utils.fold_lefti(
+    (offset, line_number, line_str) =>
+      if (position.line == line_number) {
+        offset + position.character;
+      } else if (line_number < position.line) {
+        // Add 1 for the newline
+        offset + String.length(line_str) + 1;
+      } else {
+        offset;
+      },
+    0,
+    lines,
+  );
+};
+let find_completable_state = (documents, uri, position: Protocol.position) => {
+  // try and find the code we are completing in the original source
+  switch (Hashtbl.find_opt(documents, uri)) {
+  | None => None
+  | Some(source_code) =>
+    // Get Document
+    let source =
+      Str.global_replace(Str.regexp({|\r\n|}), {|\n|}, source_code);
+    let offset = convert_position_to_offset(source, position);
+    let source_chars =
+      List.rev(
+        String_utils.explode(String_utils.slice(~last=offset, source)),
+      );
```

Comment:
This is really expensive too. I imagine it's cheaper just to check string prefixes directly.